tests/build/run_cython_not_in_cwd.srctree

```diff
@@ -0,0 +1,56 @@
+
+PYTHON check_paths_from_cython.py
+
+######## src/foo/__init__.py ########
+
+# foo is a package and foo.bar is the module to build
+
+######## src/foo/bar.pyx ########
+
+def func(x):
+    return 2*x
+
+######## build/src/foo/placeholder.txt ########
+
+# Build outputs go in the build directory and cython is run from here.
+
+######## check_paths_from_cython.py ########
+
+import sys
+import subprocess
+
+python = sys.executable
+
+def cython(cmdstr, cwd=None):
+    argv = [sys.executable, '-m', 'cython'] + cmdstr.split()
+    subprocess.run(argv, check=True, cwd=cwd)
+
+def grep(fname, pattern):
+    with open(fname) as fin:
+        return [line for line in fin if pattern in line]
+
+cython('-3 src/foo/bar.pyx -o build/src/foo/bar1.pyx.c')
+cython('-3 ../src/foo/bar.pyx -o src/foo/bar2.pyx.c', cwd='build')
+
+res1 = grep('build/src/foo/bar1.pyx.c', 'bar.pyx')
+res2 = grep('build/src/foo/bar2.pyx.c', 'bar.pyx')
+
+assert res1 == [
+    # Running cython on a file in CWD uses some paths relative to CWD.
+    '  "src/foo/bar.pyx",\n',
+    'static const char __pyx_k_src_foo_bar_pyx[] = "src/foo/bar.pyx";\n',
+    '/* "foo/bar.pyx":2\n',
+    '  /* "foo/bar.pyx":3\n',
+    '  /* "foo/bar.pyx":2\n',
+    '  /* "foo/bar.pyx":2\n'
+], str(res1)
+
+assert res2 == [
+    # When not in CWD all paths are relative to sys.path.
+    '  "foo/bar.pyx",\n',
+    'static const char __pyx_k_foo_bar_pyx[] = "foo/bar.pyx";\n',
+    '/* "foo/bar.pyx":2\n',
+    '  /* "foo/bar.pyx":3\n',
+    '  /* "foo/bar.pyx":2\n',
+    '  /* "foo/bar.pyx":2\n',
+], str(res2)
```

Comment:
The first case here (`res1`) is the typical setuptools case. This PR is intended not to change what happens in that case when cython is run from within cwd. Cython's coverage plugin already works in that case.

The second case here is what happens when building with meson or cmake. The behaviour on master would be to output a mixture of basename `bar.pyx` and absolute `/home/.../bar.pyx` paths. The change here is that all paths are `foo/bar.pyx` which is relative to sys.path in the sense that the built module would be imported as `foo.bar` rather than `src.foo.bar`.

The test case shows that actually the first case (`res1`) is a bit inconsistent. I actually think that it would be better if the second case `res2` was always the output generated by Cython regardless of CWD and that in fact Cython should not use CWD for anything besides parsing relative paths at the CLI. Changing that first case `res1` to be like `res2` would likely break some things including possibly Cython's current coverage plugin when using setuptools.

Using import-relative paths (`foo/bar.pyx`) is the most robust choice given that at cythonisation time there is no way to know where the built module will end up (e.g. site-packages, a different machine etc) or what the CWD will be when the module is actually imported. All we can really say is that the module will end up being on `sys.path` under a directory `foo` so it makes sense to use `foo/bar.pyx` as the canonical path that is baked into the source and used for trace events. The coverage plugin is then responsible for knowing that in the source tree `foo/bar.pyx` is a path relative to `src` so that from project root the relative path would be `src/foo/bar.pyx`.

This PR only changes the second case `res2` which on master is broken at least as far measuring coverage goes. Hopefully then that is only an improvement without breaking something.